### DISCUSSION TITLE
How to properly handle files (file://)

### DISCUSSION THREAD

**Discussion by marindevs** [TypeFox/monaco-languageclient#545] [Q&A]
I know from some other issues posted, that for some Language Servers, the file:// protocol will be more suitable to handle the files. For Python for example, how can I use local files and have them to work correctly with Monaco?

So I create the editor passing the path ‘C:\py_workspace\main.py’ and inside this path I have a definition.py.
I can import this module, I can use it's contents, but if I go to the Peek -> Definitions or References for members of this module, the selection is empty and I see an error in the console.

```
Uncaught Error: Unable to read file 'c:\py_workspace\globals.py' (Error: Unable to resolve nonexistent file 'c:\py_workspace\globals.py')

Error: Unable to read file 'c:\py_workspace\globals.py' (Error: Unable to resolve nonexistent file 'c:\py_workspace\globals.py')
    at MemoryFileService2.restoreReadError (fileService.js:406:18)
    at MemoryFileService2.doReadFileStream (fileService.js:390:24)
    at async BrowserTextFileService2.doRead (textFileService.js:157:28)
    at async BrowserTextFileService2.readStream (textFileService.js:139:41)
    at async _a17.resolveFromFile (textFileEditorModel.js:286:29)
    at async _a17.resolve (textFileEditorModel.js:187:9)
    at async TextFileEditorModelManager2.doResolve (textFileEditorModelManager.js:280:13)
    at async ResourceModelCollection2.doCreateReferencedObject (textModelResolverService.js:48:27)
    at errors.js:15:27
```

The same happens for built ins, event though this file exists, the Editor will throw a similar error, saying that it was unable to read the file.

```
Uncaught (in promise) Error: Unable to read file 'd:\PYTHON_PROJECTS\monaco-languageclient\node_modules\pyright\dist\typeshed-fallback\stdlib\math.pyi' (Error: Unable to resolve nonexistent file 'd:\PYTHON_PROJECTS\monaco-languageclient\node_modules\pyright\dist\typeshed-fallback\stdlib\math.pyi')
    at MemoryFileService2.restoreReadError (fileService.js:406:18)
    ... (Same as the other one)
```
Is there any way to make this work? I know there is some issues with accessing local files inside a webpage. Is it some sort of permission that I need to grant for the file to correctly load? In this case the Editor will be hosted by a WebView2, running as a native windows app.

Thanks!

**Comment by CGNonofr** [comment]
Most language servers are expecting to have access to the same filesystem as vscode.

With that lib, the filesystem is virtual, so in order for everything to work properly, you'll need to "synchronize" the client filesystem with the server.

It can be done by hands by registering a filesystem overlay that will send a request to your server by any protocol you want every time a file changes. (cons: you need to code and maintain something for it)

It can also be handled by the vscode server that is pluggable to that library since the last version, which will make the remote filesystem available on the client (cons: you need to be able to start a server with a websocket on your server)
